tests/support/mod.rs

```diff
@@ -197,39 +206,62 @@ impl RlsHandle {
 
         let start = Instant::now();
 
-        while start.elapsed() < timeout {
+        while self.within_timeout(start, timeout) {
             if let Some(ecode) = self
                 .child
                 .try_wait()
                 .expect("failed to wait on child rls process")
             {
-                assert!(ecode.success());
+                assert!(ecode.success(), "rls exit code {}", ecode);
+                // wait for stdout thread to finish to avoid races
+                while Arc::strong_count(&self.stdout) > 1 {
+                    assert!(self.within_timeout(start, timeout));
+                    thread::yield_now();
+                }
                 return self.stdout();
             }
         }
         panic!("Timed out shutting down rls");
     }
+
+    /// Uses the `call_start` or last stdout write instant, whichever is later,
+    /// to measure if the timeout has been passed.
+    ///
+    /// Also uses `timeout * 10` from the `call_start` as an absolute limit.
+    fn within_timeout(&self, call_start: Instant, timeout: Duration) -> bool {
```

Comment:
I'm sorry for nitpicking but despite the simple logic I somehow had a problem quickly grokking the code. Maybe we could extract `call_start.elapsed` and see if it helps?
Like:
```rust
let call_elapsed = call_start.elapsed();
let stdout_elapsed = self.stdout.lock().unwrap().1.elapsed();

call_elapsed.min(stdout_elapsed) < timeout && call_elapsed < timeout * 10
```

Reply:
looks good

Reply:
Thank you!